lib/clients/AliClient.py

```diff
@@ -69,4 +154,397 @@ def _download_from_blobstore(self, blob_to_download_name, blob_download_target_p
+
+    def _create_snapshot(self, volume_id, description='Service-Fabrik: Automated backup'):
+        log_prefix = '[SNAPSHOT] [CREATE]'
+        snapshot = None
+        snapshot_creation_operation = None
+        snapshot_name = self.generate_name_by_prefix(self.SNAPSHOT_PREFIX)
+        try:
+            snapshot_req_params = {
+                'DiskId': volume_id,
+                'SnapshotName': snapshot_name,
+                'Description': description
+            }
+            self.logger.info('{} START for volume id {} with tags {} and snapshot name {}'.format(
+                log_prefix, volume_id, self.tags, snapshot_name))
+            snpshot_creation_request = self._get_common_request('CreateSnapshot', snapshot_req_params, self.tags)
+            snapshot_creation_operation = self.compute_client.do_action_with_exception(snpshot_creation_request)
+            snapshot_details_json = json.loads(snapshot_creation_operation.decode('utf-8'))
+            snapshot_id = snapshot_details_json['SnapshotId']
+            self._wait('Waiting for snapshot {} to get ready...'.format(snapshot_name),
+                       (lambda snapshot_id: self._is_snapshot_ready(snapshot_id)),
+                       None,
+                       snapshot_id)
+
+            snapshot = self._get_snapshot_with_exception(snapshot_id)
+            if snapshot.status == 'accomplished':
+                self._add_snapshot(snapshot.id)
+                self.output_json['snapshotId'] = snapshot.id
+                self.logger.info('{} SUCCESS: snapshot-id={}, volume-id={}, status={} with tags {}'.format(
+                    log_prefix, snapshot.id, volume_id, snapshot.status, self.tags))
+            else:
+                message = '{} ERROR: snapshot-id={} status={}'.format(
+                    log_prefix, snapshot_name, snapshot.status)
+                raise Exception(message)
+        except Exception as error:
+            message = '{} ERROR: volume-id={} and tags={}\n{}'.format(
+                log_prefix, volume_id, self.tags, error)
+            self.logger.error(message)
+            if snapshot or snapshot_creation_operation:
+                self.delete_snapshot(snapshot_id)
+                snapshot = None
+            raise Exception(message)
+
+        return snapshot
+
+    def _get_snapshot_list(self, snapshot_id):
+        region_id = self.__aliCredentials['region_name']
+        get_snapshot_req_params = {
+            'PageSize' : 10,
+            'RegionId' : region_id,
+            'SnapshotIds' : [snapshot_id]
+        }
+        get_snapshot_request = self._get_common_request('DescribeSnapshots', get_snapshot_req_params)
+        snapshot_details = self.compute_client.do_action_with_exception(get_snapshot_request)
+        snapshot_details_json = json.loads(snapshot_details.decode('utf-8'))
+        return snapshot_details_json['Snapshots']['Snapshot']
+    
+    def _get_snapshot_with_exception(self, snapshot_id):
```

Comment:
minor thing, can we rename it to something better ?

Reply:
Yes, method with same functionality can be merged. I would suggest remove the method. Use the 'get_snapshot'. In caller wherever required e.g. [here](https://github.com/cloudfoundry-incubator/service-fabrik-backup-restore/pull/57/files#diff-fe5e9623c538d6846d10874ca9426104R181) use conditional coding by checking None/Empty

Reply:
I'd want to populate the exact exception coming from above hence using the same method doesn't work. But what we can do is add retry method in base class and through None there with retries. It would mean that we retry on exception which we don't do for some apis. Does that work?

Reply:
Going ahead w/o changing implementation of other clients. For AliClient I'm keeping retries